sympy/core/tests/test_power.py

```diff
@@ -361,7 +361,8 @@ def test_issue_7638():
     e = 1/(1 - sqrt(2))
     assert sqrt(e) == I/sqrt(-1 + sqrt(2))
     assert e**-S.Half == -I*sqrt(-1 + sqrt(2))
-    assert sqrt((cos(1)**2 + sin(1)**2 - 1)**(3 + I)).exp == S.Half
+    assert sqrt((cos(1)**2 + sin(1)**2 - 1)**(3 + I)).exp in [S.Half,
+                                                              S(3)/2 + I/2]
```

Comment:
@smichr I had to update this. Not sure what the correct value should be though. The reason it changes is that 
`re(cos(1)**2 + sin(1)**2 - 1)` happens to evaluate to `-3.78182804184504e-124` so the piecewise returns `pi`, where it should have returned NaN. I assume something similar happens in the original test, but `-S.ImaginaryUnit*log((x + S.ImaginaryUnit*y)/sqrt(x**2 + y**2))` for `y = 0` and `s = cos(1)**2 + sin(1)**2 - 1` returns 
![image](https://user-images.githubusercontent.com/8114497/61995675-bed3b900-b08b-11e9-8717-365a09c749fb.png)

which probably also evaluates in a similar way.

Reply:
I can't see in the diff where this evaluates or how evaluation is used to make a decision. This test is likely intending to deal with `sqrt(x**(3+I))` where `x` is a number that is not known to be zero or not -- it is undecidable without simplification. If Piecewise is using evaluation to make a decision, it should only do so if `x.is_comparable`...or something to that effect.

Reply:
It seems like `re` (or the comparison, but just doing  `re(cos(1)**2 + sin(1)**2 - 1)` leads to a float result) triggers the evaluation. I simply enter the Piecewise (if y.is_zero and x.is_number).

Since the expression should be zero, I guess it is hard to say what the argument really should be? (And it isn't really obvious how the test expression comes to atan2, but it does, multiple times...)

Reply:
Numerical evaluation is used in the assumptions for inequalities sometimes. See the `_n2` function used in `Expr.__lt__` etc.

Reply:
Although it looks like the `_n2` function is called here and returns None because:
```
In [1]: (cos(1)**2 + sin(1)**2 - 1).is_comparable
Out[1]: False
```

Reply:
Ah, so it is `is_comparable` one should use? However, I dread that it may ruin the fix as `S.ImaginaryUnit.is_comparable` is `False` and therefore there is no way to tell if one can apply the new evaluation code or not (and it is needed to avoid the` RecursionError`). (Note that the actual comparisons only involves real numbers, although the expression that is available, without applying re or im which will apparently lead to evaluation in some cases, will be complex.)